src/imports/background/import-item-creation.js

```diff
@@ -0,0 +1,195 @@
+     * @param {Array<any>} items Array of items with `url` to filter on.
+     * @param {(item: any) => any} transform Opt. transformformation fn turning current iterm into import item structure.
+     * @param {(url: string) => bool} alreadyExists Opt. checker function to check against existing data.
+     * @return {Map<string, any>} Filtered version of `items` array made into a Map of encoded URL strings to import items.
+     */
+    return (items, transform = f => f, alreadyExists) =>
+        items.reduce((accItems, currItem) => {
+            if (!isLoggable(currItem) || !isNotBlacklisted(currItem)) {
+                return accItems
+            }
+
+            try {
+                // Augment the item with encoded URL for DB check and item map
+                const encodedUrl = encodeUrl(currItem.url, false)
+
+                // Filter out items already existing in DB, if specified
+                if (alreadyExists == null || !alreadyExists(encodedUrl)) {
+                    accItems.set(encodedUrl, transform(currItem))
+                }
+                return accItems
+            } catch (error) {
+                // URI malformed; don't add to list
+                return accItems
+            }
+        }, new Map())
+}
+
+/**
+ * @param {number} [maxResults=9999999] The max amount of items to grab from browser.
+ * @param {number} [startTime=0] The time to start search from for browser history.
+ * @returns {Array<browser.history.HistoryItem>} All history items in browser filtered by URL.
+ */
+const getHistoryItems = (maxResults = 9999999, startTime = 0) =>
+    browser.history.search({
+        text: '',
+        maxResults,
+        startTime,
+    })
+
+/**
+ * @param {number} [maxResults=100000] The max amount of items to grab from browser.
+ * @returns {Array<browser.bookmarks.BookmarkTreeNode>} All bookmark items in browser filtered by URL.
+ */
+const getBookmarkItems = (maxResults = 100000) =>
+    browser.bookmarks.getRecent(maxResults)
+
+export async function getOldExtItems() {
+    let bookmarkUrls = new Set()
+    const {
+        [OLD_EXT_KEYS.INDEX]: index,
+        [OLD_EXT_KEYS.BOOKMARKS]: bookmarks,
+    } = await browser.storage.local.get({
+        [OLD_EXT_KEYS.INDEX]: { index: [] },
+        [OLD_EXT_KEYS.BOOKMARKS]: '[]',
+    })
+
+    if (typeof bookmarks === 'string') {
+        try {
+            bookmarkUrls = new Set(JSON.parse(bookmarks).map(bm => bm.url))
+        } catch (error) {}
+    }
+
+    const transform = transformOldExtDataToImportItem(bookmarkUrls)
+    const importItems = []
+
+    // Only attempt page data conversion if index + bookmark storage values are correct types
+    if (index && index.index instanceof Array) {
+        // NOTE: There is a bug with old ext index sometimes repeating the index keys
+        //  (eg.I indexed 400 pages, index contained 43 million)
+        const indexSet = new Set(index.index)
+
+        for (const keyChunk of chunk(200)([...indexSet])) {
```

Comment:
I have no idea of the performance of chunk and the spread operator (I assume quite low compared to a forloop ending at 200.

Reply:
Reading the core more I'm not sure I understand the reason for this chunking in the first place, and what ```browser.storage.local.get(keyChunk)``` contains?

Reply:
This is the old extension data. Each page/visit was stored as its own key in local storage. There is a special key `index` which is an array of all the other keys, acting as the reverse index. 

So here we're grabbing `index` array, breaking it into chunks, and then accessing that chunk of 200 docs from local storage, transforming it to new structure, before moving onto next chunk. Currently splitting it into chunks with the thoughts (both relative to the # of old ext docs):
 - put a constant cap on space
 - give a `N / #chunks` speedup in time (compared to `N` if we access one-at-a-time)

Could be replicated with old-style for-loop, although that would just be another sub-optimization, right? Might as well give it a try

Reply:
if this is a one-off from the old index to the new one probably is not worth spending more time that you did